WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -278,6 +278,20 @@ def update_events
     end
   end
 
+  def update_wcif_events
+    @competition = competition_from_params
+    wcif_events = params["_json"].map { |wcif_event| wcif_event.permit!.to_h } #<<< TODO: validate that the incoming stuff is actually wcif >>>
```

Comment:
What absolutely has to be guarded against is SQL injection attacks or anything like that. Maybe that happens automatically because Rails is great, IDK.

As for validating the document, it can be a huge time and effort sink. Possibly also duplicating things implemented for other paths. If we can get away with just relying on the validations in the model objects, that would be really great.

Of course, they are written assuming input comes from forms that the app itself creates. If nothing else, such error messages can be profoundly unhelpful to someone trying to hand craft a valid WCIF document. Getting back controller exceptions from code that expects a well formed document sucks.

I could imagine that something that validates the _structure_ of the document, but not the data could be a good bang-for-buck point. That is, the "tree" should have these "directory" nodes to even have a chance to be valid, but leave validating the individual fields to the model validators.

Looking quickly at the JSON schema thing, it's hard to say how the cost of it would compare to the benefits.

What is it that makes you convinced validation is needed?